<launch>

  <arg name="launch_robot" default="true"/>
  <arg name="launch_rviz" default="true"/>
  <arg name="launch_infinitam" default="true"/>
  <arg name="launch_planner" default="true"/>
  <arg name="enable_guess_from_tf_tree" default="true"/>

  <arg name="launch_delay" default="0"/>

  <arg name="enable_debug" default="false"/>
  <arg name="launch_manager"     default="true"/>
  <arg name="manager" default="local_planner"/>

  <arg name="launch_prefix" default="gdb -ex run --args" if="$(eval arg('enable_debug') and arg('launch_manager'))"/>
  <arg name="launch_prefix" default="" unless="$(eval arg('enable_debug') and arg('launch_manager'))"/>

  <node pkg="nodelet" type="nodelet" name="$(arg manager)" output="screen" respawn="true" launch-prefix="$(arg launch_prefix)"
    args="manager" if="$(arg launch_manager)"/>

  <arg name="basepath" value="$(find mochapc)"/>
  <include file="$(find carplanner_msgs)/launch/demo.launch" if="$(arg launch_planner)">
    <arg name="params_file" value="$(arg basepath)/husky_params.csv"/>
    <arg name="terrain_mesh_file" value="$(arg basepath)/disk-ascii.ply"/>
    <arg name="car_mesh_file" value="$(arg basepath)/herbie/herbie.blend"/>
    <arg name="wheel_mesh_file" value="$(arg basepath)/herbie/wheel.blend"/>
    <arg name="map_frame" value="H01/map"/>
    <arg name="base_link_frame" value="H01/base_link"/>

    <arg name="launch_infinitam" default="false"/>
    <arg name="launch_delay" value="$(eval launch_delay+0)"/>
    <arg name="enable_debug" value="$(arg enable_debug)"/>
    <arg name="launch_manager" value="false"/>
    <arg name="manager" value="$(arg manager)"/>
  </include>

  <!-- <node pkg="tf2_ros" type="static_transform_publisher" name="T_camdown_optical" args="0 0 0 -1.57 0 -1.57 X1/base_link/camera_down X1/base_link/camera_down_optical"/> -->
  <!-- <include file="$(find infinitam)/launch/subt_sim_demo.launch">
  </include> -->

  <include file="$(find infinitam)/launch/subt_demo.launch" if="$(arg launch_infinitam)">
    <arg name="launch_robot" value="false"/>
    <arg name="launch_rviz" value="false"/>
    <arg name="enable_guess_from_tf_tree" value="$(arg enable_guess_from_tf_tree)"/>

    <arg name="launch_delay" value="$(eval launch_delay+10)"/>
    <arg name="enable_debug" value="$(arg enable_debug)"/>
    <arg name="launch_manager" value="false"/>
    <arg name="manager" value="$(arg manager)"/>
  </include>

  <arg name="wg_delay" default="$(eval launch_delay+0)"/>
  <node pkg="carplanner_msgs" type="waypoint_generator" name="local_goal2local_waypoints" output="log" launch-prefix="bash -c 'sleep $(arg wg_delay); $0 $@'">
    <param name="odom_topic" value="H01/odometry"/>
    <param name="goal_topic" value="planner/goal"/>
    <param name="waypoints_topic" value="planner/input_waypoints"/>
    <!-- <param name="map_frame_id" value="X1/map"/>
    <param name="base_frame_id" value="X1/base_link"/> -->
    <param name="rate" value="100"/>
  </node>

  <node pkg="topic_tools" type="relay" name="local_goal_relay2" args="/move_base_simple/goal planner/goal"/>
  <node pkg="topic_tools" type="relay" name="local_goal_relay" args="/rviz/nav_goal_2d planner/goal"/>
  <!-- <node pkg="topic_tools" type="relay" name="global_path_relay" args="/H01/ma_goal_path planner/global_path"/> -->
  <node pkg="topic_tools" type="relay" name="local_path_relay" args="planner/actual_traj /X1/local_path"/>

 <arg name="og_delay" default="$(eval launch_delay+0)"/>
 <node pkg="carplanner_msgs" type="tf2odom" name="tf2odom" output="log" launch-prefix="bash -c 'sleep $(arg og_delay); $0 $@'">
    <param name="parent_frame" value="H01/map"/>
    <param name="child_frame" value="H01/base_link"/>
    <param name="odom_topic" value="H01/odom"/>
    <param name="rate" value="100"/>
  </node> 

  <arg name="gg_delay" default="$(eval launch_delay+0)"/>
  <node pkg="carplanner_msgs" type="goal_generator" name="global_path2local_goal" output="log" launch-prefix="bash -c 'sleep $(arg gg_delay); $0 $@'">
    <param name="path_topic" value="planner/global_path"/>
    <param name="goal_topic" value="planner/goal"/>
    <param name="odom_topic" value="/H01/odometry"/>
    <param name="path_intercept_length" value="3.0"/>
    <param name="rate" value="100"/>
  </node>

  <include file="$(find marble_common)/launch/common/main.launch" if="$(arg launch_robot)">
    <arg name="vehicle_name"              value="H01"/>
    <arg name="vehicle_type"              value="Husky"/>
    <arg name="enable_localization"       value="1"/>
    <arg name="enable_gate_localization"  value="0"/>
    <arg name="enable_mapping"            value="1"/>
    <arg name="enable_multi_agent"        value="1"/>
    <arg name="enable_multimaster"        value="0"/>
    <arg name="enable_planning"           value="1"/>
    <arg name="enable_control"            value="1"/>
    <arg name="enable_marble_interface"   value="1"/>
    <arg name="enable_teleop"             value="1"/>
    <arg name="enable_rviz"               value="0"/>
    <arg name="enable_udp"                value="1"/>
    <arg name="enable_artifact_detection" value="0"/>
    <arg name="localization_type"         value="0"/>
    <arg name="use_secondary"             value="0"/>
    <arg name="motion_on_startup"         value="0"/>
  </include>

  <node pkg="rviz" type="rviz" name="subt_tap_rviz" if="$(arg launch_rviz)"
    args="-d $(env HOME)/.rviz/subt_tap.rviz"/>

</launch>
